### PR TITLE
JAMES-2813 Use JsonGenericSerializer to serialize Tasks

### PR DESCRIPTION
**Pull request by blackheaven**

### CODE REVIEW

**Comment by Arsnael** [comment]
```
[a1848554f4f3f4e57a69f7d2d7b724f9ea377fb3] [ERROR] /james-project/mailbox/tools/indexer/src/main/java/org/apache/mailbox/tools/indexer/dto/SingleMailboxReindexingTaskDTO.java:[22,45] package org.apache.james.server.task.json.dto does not exist
[a1848554f4f3f4e57a69f7d2d7b724f9ea377fb3] [ERROR] /james-project/mailbox/tools/indexer/src/main/java/org/apache/mailbox/tools/indexer/dto/SingleMailboxReindexingTaskDTO.java:[26,56] cannot find symbol
[a1848554f4f3f4e57a69f7d2d7b724f9ea377fb3]   symbol: class TaskDTO
[a1848554f4f3f4e57a69f7d2d7b724f9ea377fb3] [ERROR] /james-project/mailbox/tools/indexer/src/main/java/org/apache/mailbox/tools/indexer/dto/SingleMailboxReindexingTaskDTO.java:[36,5] method does not override or implement a method from a supertype
```

File: event-sourcing/event-store-cassandra/src/main/java/org/apache/james/eventsourcing/eventstore/cassandra/JsonEventSerializer.java [hunk 1]
```diff
@@ -34,6 +34,13 @@
 
 public class JsonEventSerializer {
 
+    public static class InvalidEventException extends RuntimeException {
+        public InvalidEventException(JsonGenericSerializer.InvalidTypeException original) {
+            super(original);
+        }
+    }
+
+
```
Comment (Arsnael) [inline event-sourcing/event-store-cassandra/src/main/java/org/apache/james/eventsourcing/eventstore/cassandra/JsonEventSerializer.java]: remove extra line break

File: event-sourcing/event-store-cassandra/src/test/java/org/apache/james/eventsourcing/eventstore/cassandra/JsonEventSerializerTest.java [hunk 2]
```diff
@@ -84,6 +86,25 @@ void deserializeShouldHandleAllKnownEvents() throws Exception {
             .isEqualTo(TEST_EVENT);
     }
 
+
+    @Test
+    void deserializeShouldThrowWhenEventWithDuplicateTypes() {
```
Comment (Arsnael) [inline event-sourcing/event-store-cassandra/src/test/java/org/apache/james/eventsourcing/eventstore/cassandra/JsonEventSerializerTest.java]: s/Duplicate/Duplicated

File: json/src/test/java/org/apache/JsonGenericSerializerTest.java [hunk 3]
```diff
@@ -49,6 +51,22 @@ void shouldDeserializeKnownType() throws Exception {
...
 
+    @Test
+    void shouldThrowWhenDeserializeEventWithMissingType() {
+        assertThatThrownBy(() -> JsonGenericSerializer.of(TestModules.FIRST_TYPE)
+            .deserialize(MISSING_TYPE_JSON))
+            .isInstanceOf(JsonGenericSerializer.InvalidTypeException.class);
+    }
+
+    @Test
+    void shouldThrowWhenDeserializeEventWithDuplicateType() {
```
Comment (Arsnael) [inline json/src/test/java/org/apache/JsonGenericSerializerTest.java]: s/DuplicateType/DuplicatedTypes

File: json/src/test/java/org/apache/dto/TestModules.java [hunk 4]
```diff
@@ -25,24 +25,26 @@
...
-                    domainObject.getTime().toString(),
-                    domainObject.getPayload()))
-            .typeName("first")
-            .withFactory(TestModule::new);
+        .forDomainObject(FirstDomainObject.class)
+        .convertToDTO(FirstDTO.class)
+        .toDomainObjectConverter(FirstDTO::toDomainObject)
+        .toDTOConverter((domainObject, typeName) -> new FirstDTO(
+            typeName,
+                domainObject.getId(),
```
Comment (Arsnael) [inline json/src/test/java/org/apache/dto/TestModules.java]: indent issue with typeName

File: mailbox/tools/indexer/src/test/java/org/apache/mailbox/tools/indexer/SingleMailboxReindexingTaskTest.java [hunk 5]
```diff
@@ -0,0 +1,21 @@
+package org.apache.mailbox.tools.indexer;
```
Comment (Arsnael) [inline mailbox/tools/indexer/src/test/java/org/apache/mailbox/tools/indexer/SingleMailboxReindexingTaskTest.java]: Apache license missing

Comment (Arsnael) [reply mailbox/tools/indexer/src/test/java/org/apache/mailbox/tools/indexer/SingleMailboxReindexingTaskTest.java]: Forget it... The file has been removed? The fact that github shows this commit being made 20h ago against 2 days for the others is fooling...

**Comment by trantienduchn** [comment]
```
[a1848554f4f3f4e57a69f7d2d7b724f9ea377fb3] [ERROR] COMPILATION ERROR : 
[a1848554f4f3f4e57a69f7d2d7b724f9ea377fb3] [INFO] -------------------------------------------------------------
[a1848554f4f3f4e57a69f7d2d7b724f9ea377fb3] [ERROR] /james-project/mailbox/tools/indexer/src/main/java/org/apache/mailbox/tools/indexer/dto/SingleMailboxReindexingTaskDTO.java:[22,45] package org.apache.james.server.task.json.dto does not exist
[a1848554f4f3f4e57a69f7d2d7b724f9ea377fb3] [ERROR] /james-project/mailbox/tools/indexer/src/main/java/org/apache/mailbox/tools/indexer/dto/SingleMailboxReindexingTaskDTO.java:[26,56] cannot find symbol
[a1848554f4f3f4e57a69f7d2d7b724f9ea377fb3]   symbol: class TaskDTO
[a1848554f4f3f4e57a69f7d2d7b724f9ea377fb3] [ERROR] /james-project/mailbox/tools/indexer/src/main/java/org/apache/mailbox/tools/indexer/dto/SingleMailboxReindexingTaskDTO.java:[36,5] method does not override or implement a method from a supertype
```

**Comment by trantienduchn** [comment]
```
[38a5e83c3d4bde66a03eeafa47582ea829556b5b] Location: /org/apache/james/imap/scripts/ValidateDeployment.test:499
[38a5e83c3d4bde66a03eeafa47582ea829556b5b] LastClientMsg: c SEARCH FLAGGED
[38a5e83c3d4bde66a03eeafa47582ea829556b5b] Expected: '\* SEARCH 1 2 3 4'
[38a5e83c3d4bde66a03eeafa47582ea829556b5b] Actual   : '* SEARCH'
```
test this please

**Comment by blackheaven** [comment]
:green_apple: before rebase

**Comment by rouazana** [comment]
```
[95829da0b50db714116b660f20a41a2993140c93] [ERROR] src/main/java/org/apache/james/task/Task.java:[26,8] (imports) UnusedImports: Unused import - org.apache.commons.lang3.NotImplementedException.
[95829da0b50db714116b660f20a41a2993140c93] [ERROR] src/main/java/org/apache/james/task/Task.java:[100] (whitespace) EmptyLineSeparator: 'METHOD_DEF' should be separated from previous statement.
```

**Comment by blackheaven** [comment]
test this please

**Comment by blackheaven** [comment]
test this please

**Comment by rouazana** [comment]
merged

**Pull request closed without merging**